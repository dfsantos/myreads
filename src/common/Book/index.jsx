import React from 'react';
import PropTypes from 'prop-types';

const propTypes = {
  title: PropTypes.string.isRequired,
};

const style = {
  width: 128,
  height: 193,
  backgroundImage:
    'url("http://books.google.com/books/content?id=uu1mC6zWNTwC&printsec=frontcover&img=1&zoom=1&imgtk=AFLRE73pGHfBNSsJG9Y8kRBpmLUft9O4BfItHioHolWNKOdLavw-SLcXADy3CPAfJ0_qMb18RmCa7Ds1cTdpM3dxAGJs8zfCfm8c6ggBIjzKT7XR5FIB53HHOhnsT7a0Cc-PpneWq9zX&source=gbs_api")',
};

function Book({ title }) {
  return (
    <div className="book">
      <div className="book-top">
        <div className="book-cover" style={style} />
        <div className="book-shelf-changer">
          <select>
            <option value="none" disabled>
              Move to...
            </option>
            <option value="currentlyReading">Currently Reading</option>
            <option value="wantToRead">Want to Read</option>
            <option value="read">Read</option>
            <option value="none">None</option>
          </select>
        </div>
      </div>
      <div className="book-title">{title}</div>
      <div className="book-authors">{[].join(', ')}</div>
    </div>
  );
}

Book.propTypes = propTypes;

export default Book;
