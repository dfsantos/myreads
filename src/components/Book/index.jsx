import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { Card, CardText, CardActions, CardMedia } from 'material-ui/Card';
import DropDownMenu from 'material-ui/DropDownMenu';
import MenuItem from 'material-ui/MenuItem';

const propTypes = {
  title: PropTypes.string.isRequired,
  authors: PropTypes.arrayOf(PropTypes.string),
  coverLink: PropTypes.string.isRequired,
  onChangeBookShelf: PropTypes.func.isRequired,
  shelf: PropTypes.string,
};

const defaultProps = {
  authors: [],
  shelf: 'none',
};

const SELECTED_VALUE = 2;

class Book extends Component {
  constructor(props) {
    super(props);
    this.state = { shelf: props.shelf, isWaitingResponse: false };
    this.onChangeBookShelf = this.onChangeBookShelf.bind(this);
  }

  componentWillReceiveProps(nextProps) {
    this.setState({ shelf: nextProps.shelf, isWaitingResponse: false });
  }

  onChangeBookShelf(...args) {
    this.setState({ isWaitingResponse: true });
    this.props.onChangeBookShelf(args[SELECTED_VALUE]);
  }

  render() {
    const { title, authors, coverLink } = this.props;
    const { shelf, isWaitingResponse } = this.state;
    const authorInfo = !authors.length ? 'Unknown Author' : authors.join(', ').trim();

    return (
      <Card className="card">
        <CardMedia>
          <div className="cover-wrapper">
            <img src={coverLink} alt={title} />
          </div>
        </CardMedia>
        <CardText>{authorInfo}</CardText>
        <CardActions>
          {isWaitingResponse && <span className="updating-message">Updating...</span>}
          {!isWaitingResponse && (
            <DropDownMenu value={shelf} onChange={this.onChangeBookShelf}>
              <MenuItem value="none" primaryText="None" />
              <MenuItem value="currentlyReading" primaryText="Currently Reading" />
              <MenuItem value="wantToRead" primaryText="Want To Read" />
              <MenuItem value="read" primaryText="Read" />
            </DropDownMenu>
          )}
        </CardActions>
      </Card>
    );
  }
}

Book.propTypes = propTypes;
Book.defaultProps = defaultProps;

export default Book;
