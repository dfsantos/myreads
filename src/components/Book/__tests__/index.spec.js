import React from 'react';
import toJson from 'enzyme-to-json';

import Book from '../index';

const TITLE = "Book's Title";

const AUTHORS = ['Author of Book'];

const onChangeBookCategoryListener = jest.fn();

it('should render without errors', () => {
  shallow(
    <Book
      title={TITLE}
      coverLink=""
      authors={AUTHORS}
      onChangeBookCategory={onChangeBookCategoryListener}
    />
  );
});

it('default snapshot', () => {
  const wrapper = shallow(
    <Book
      title={TITLE}
      coverLink=""
      authors={AUTHORS}
      onChangeBookCategory={onChangeBookCategoryListener}
    />
  );
  expect(toJson(wrapper)).toMatchSnapshot();
});

it('on select change should call onChangeBookCategoryListener with selected value', () => {
  const SELECTED_VALUE = 'read';
  const event = { target: { value: SELECTED_VALUE } };
  const wrapper = shallow(
    <Book
      title={TITLE}
      coverLink=""
      authors={AUTHORS}
      onChangeBookCategory={onChangeBookCategoryListener}
    />
  );
  const select = wrapper.find('select');

  select.simulate('change', event);

  expect(onChangeBookCategoryListener).toHaveBeenCalledWith(SELECTED_VALUE);
});
