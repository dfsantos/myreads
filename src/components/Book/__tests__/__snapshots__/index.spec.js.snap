// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default snapshot 1`] = `
<Card
  className="card"
  expandable={false}
  expanded={null}
  initiallyExpanded={false}
>
  <CardMedia>
    <div
      className="cover-wrapper"
    >
      <img
        alt="Book's Title"
        src=""
      />
    </div>
  </CardMedia>
  <CardText>
    Author of Book
  </CardText>
  <CardActions>
    <DropDownMenu
      anchorOrigin={
        Object {
          "horizontal": "left",
          "vertical": "top",
        }
      }
      animated={true}
      autoWidth={true}
      disabled={false}
      iconButton={<NavigationArrowDropDown />}
      maxHeight={500}
      multiple={false}
      onChange={[Function]}
      openImmediately={false}
      value="none"
    >
      <MenuItem
        anchorOrigin={
          Object {
            "horizontal": "right",
            "vertical": "top",
          }
        }
        checked={false}
        desktop={false}
        disabled={false}
        focusState="none"
        insetChildren={false}
        primaryText="None"
        targetOrigin={
          Object {
            "horizontal": "left",
            "vertical": "top",
          }
        }
        value="none"
      />
      <MenuItem
        anchorOrigin={
          Object {
            "horizontal": "right",
            "vertical": "top",
          }
        }
        checked={false}
        desktop={false}
        disabled={false}
        focusState="none"
        insetChildren={false}
        primaryText="Currently Reading"
        targetOrigin={
          Object {
            "horizontal": "left",
            "vertical": "top",
          }
        }
        value="currentlyReading"
      />
      <MenuItem
        anchorOrigin={
          Object {
            "horizontal": "right",
            "vertical": "top",
          }
        }
        checked={false}
        desktop={false}
        disabled={false}
        focusState="none"
        insetChildren={false}
        primaryText="Want To Read"
        targetOrigin={
          Object {
            "horizontal": "left",
            "vertical": "top",
          }
        }
        value="wantToRead"
      />
      <MenuItem
        anchorOrigin={
          Object {
            "horizontal": "right",
            "vertical": "top",
          }
        }
        checked={false}
        desktop={false}
        disabled={false}
        focusState="none"
        insetChildren={false}
        primaryText="Read"
        targetOrigin={
          Object {
            "horizontal": "left",
            "vertical": "top",
          }
        }
        value="read"
      />
    </DropDownMenu>
  </CardActions>
</Card>
`;

exports[`snapshot without author 1`] = `
<Card
  className="card"
  expandable={false}
  expanded={null}
  initiallyExpanded={false}
>
  <CardMedia>
    <div
      className="cover-wrapper"
    >
      <img
        alt="Book's Title"
        src=""
      />
    </div>
  </CardMedia>
  <CardText>
    Unknown Author
  </CardText>
  <CardActions>
    <DropDownMenu
      anchorOrigin={
        Object {
          "horizontal": "left",
          "vertical": "top",
        }
      }
      animated={true}
      autoWidth={true}
      disabled={false}
      iconButton={<NavigationArrowDropDown />}
      maxHeight={500}
      multiple={false}
      onChange={[Function]}
      openImmediately={false}
      value="none"
    >
      <MenuItem
        anchorOrigin={
          Object {
            "horizontal": "right",
            "vertical": "top",
          }
        }
        checked={false}
        desktop={false}
        disabled={false}
        focusState="none"
        insetChildren={false}
        primaryText="None"
        targetOrigin={
          Object {
            "horizontal": "left",
            "vertical": "top",
          }
        }
        value="none"
      />
      <MenuItem
        anchorOrigin={
          Object {
            "horizontal": "right",
            "vertical": "top",
          }
        }
        checked={false}
        desktop={false}
        disabled={false}
        focusState="none"
        insetChildren={false}
        primaryText="Currently Reading"
        targetOrigin={
          Object {
            "horizontal": "left",
            "vertical": "top",
          }
        }
        value="currentlyReading"
      />
      <MenuItem
        anchorOrigin={
          Object {
            "horizontal": "right",
            "vertical": "top",
          }
        }
        checked={false}
        desktop={false}
        disabled={false}
        focusState="none"
        insetChildren={false}
        primaryText="Want To Read"
        targetOrigin={
          Object {
            "horizontal": "left",
            "vertical": "top",
          }
        }
        value="wantToRead"
      />
      <MenuItem
        anchorOrigin={
          Object {
            "horizontal": "right",
            "vertical": "top",
          }
        }
        checked={false}
        desktop={false}
        disabled={false}
        focusState="none"
        insetChildren={false}
        primaryText="Read"
        targetOrigin={
          Object {
            "horizontal": "left",
            "vertical": "top",
          }
        }
        value="read"
      />
    </DropDownMenu>
  </CardActions>
</Card>
`;
